advisor/src/main/kotlin/advisors/Osv.kt

```diff
@@ -0,0 +1,190 @@
+ * An advice provider that obtains vulnerability from OSV.dev https://osv.dev/.
+ */
+class Osv(name: String, advisorConfiguration: AdvisorConfiguration) : AdviceProvider(name) {
+    class Factory : AbstractAdviceProviderFactory<Osv>("Osv") {
+        override fun create(config: AdvisorConfiguration) = Osv(providerName, config)
+    }
+
+    override val details: AdvisorDetails = AdvisorDetails(providerName, enumSetOf(AdvisorCapability.VULNERABILITIES))
+
+    private val service = OsvService(
+        serverUrl = advisorConfiguration.osv?.serverUrl ?: OsvApiClient.SERVER_URL_PRODUCTION,
+        httpClient = OkHttpClientHelper.buildClient()
+    )
+
+    override suspend fun retrievePackageFindings(packages: List<Package>): Map<Package, List<AdvisorResult>> {
+        val startTime = Instant.now()
+
+        val vulnerabilitiesForPackage = getVulnerabilitiesForPackage(packages)
+
+        return packages.associateWith { pkg ->
+            val vulnerabilities = vulnerabilitiesForPackage[pkg.id].orEmpty().map { it.toOrtVulnerability() }
+
+            if (vulnerabilities.isNotEmpty()) {
+                listOf(
+                    AdvisorResult(
+                        advisor = details,
+                        summary = AdvisorSummary(
+                            startTime = startTime,
+                            endTime = Instant.now()
+                        ),
+                        vulnerabilities = vulnerabilities
+                    )
+                )
+            } else {
+                emptyList()
+            }
+        }
+    }
+
+    private fun getVulnerabilitiesForPackage(packages: List<Package>): Map<Identifier, List<Vulnerability>> {
+        val vulnerabilityIdsForPackageId = getVulnerabilityIdsForPackages(packages)
+        val allVulnerabilityIds = vulnerabilityIdsForPackageId.values.flatten().toSet()
+        val vulnerabilityForId = getVulnerabilitiesForIds(allVulnerabilityIds).associateBy { it.id }
+
+        return packages.associate { pkg ->
+            pkg.id to vulnerabilityIdsForPackageId[pkg.id].orEmpty().map { vulnerabilityForId.getValue(it) }
+        }
+    }
+
+    private fun getVulnerabilityIdsForPackages(packages: List<Package>): Map<Identifier, List<String>> {
+        val requests = packages.mapNotNull { pkg ->
+            val request = createRequest(pkg) ?: return@mapNotNull null
+            pkg to request
+        }
+
+        val result = service.getVulnerabilityIdsForPackages(requests.map { it.second })
+        if (result.isFailure) {
+            log.error {
+                "Requesting vulnerabilities IDs for packages failed: ${result.exceptionOrNull()!!.collectMessages()}"
+            }
+        }
+
+        return buildMap {
+            result.getOrNull().orEmpty().forEachIndexed { i, vulnerabilities ->
+                put(requests[i].first.id, vulnerabilities)
+            }
+
+            packages.forEach { pkg ->
+                putIfAbsent(pkg.id, emptyList())
+            }
```

Comment:
Is this the code that requires you to filter out packages with a empty list of vulnerabilities at another place? I can't currently find that code anymore, but I remember to have commented on it during a previous review.

Reply:
@oheger-bosch, do you remember whether returning an empty list of defects / vulnerabilities for unaffected packages was a deliberate design decision, or is it an oversight and should we simply have no entries for such packages (which is what I'd personally strongly prefer)?

Reply:
FYI @sschuberth: I just changed it according to your ask. For the private function this is very minor as visibility scope is very small, so I just changed it to shortcut the discussion.

Reply:
> @oheger-bosch, do you remember whether returning an empty list of defects / vulnerabilities for unaffected packages was a deliberate design decision, or is it an oversight and should we simply have no entries for such packages (which is what I'd personally strongly prefer)?

As far as I remember, it was a deliberate decision. As the result is a `Map` keyed by `Package`, the thinking was that there should be an entry for each package that was passed as input.